In [52]:
# importing important libraries
import mnist
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression
import warnings
warnings.filterwarnings(action="ignore")

In [53]:
# taking the test and train images(independent) and labels(target)
train_images = mnist.train_images()
train_labels = mnist.train_labels()
test_images = mnist.test_images()
test_labels = mnist.test_labels()

In [18]:
#checking the shape of train images
train_images.shape

(60000, 28, 28)

In [21]:
# flatten the images fron 2d array to 1 d array
# 2d array => 28,28 ------> 1d array=>784
train_image = train_images.reshape(-1,784)
test_image = test_images.reshape(-1,784)

In [23]:
# using the standar scaler to put the value in a certain range
from sklearn.preprocessing import StandardScaler
scalar = StandardScaler()
X_train = scalar.fit_transform(train_image)

In [25]:
scalar1 = StandardScaler()
X_test = scalar.fit_transform(test_image)

In [26]:
#PCA
from sklearn.decomposition import PCA
pca = PCA(0.95)
pca.fit(X_train)

PCA(copy=True, iterated_power='auto', n_components=0.95, random_state=None,
  svd_solver='auto', tol=0.0, whiten=False)

In [27]:
pca.n_components_

331

as you can see that PCA is  the dimension reduce technique and from the above code it reduces the column from 786 to 331
now when you build your model the model will work with 331 independet varibale to fit the model..which will reduce the time and increases the accuracy

In [28]:
X_train_pca = pca.transform(X_train)
X_test_pca = pca.transform(X_test)

In [29]:
#building the model
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(solver="lbfgs")
model.fit(X_train_pca,train_labels)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='lbfgs',
          tol=0.0001, verbose=0, warm_start=False)

In [30]:
# predicting the value on unseen data
y_pred = model.predict(X_test_pca)

In [40]:
#accuracy
from sklearn.metrics import accuracy_score
accuracy_score(y_pred,test_labels)

0.9213